### PR TITLE
[mlir][Arm] Fix invalid rewrite pattern API violations

### PR DESCRIPTION
**Pull request by matthias-springer**
This commit fixes rewrite pattern API violations:
* Rewrite pattern must return "failure" if the IR was not modified.
* In-place op modifications must be communicated to the rewriter (`updateRootInPlace`).

This commit fixes `test/Dialect/ArmSVE/legalize-vector-storage.mlir`, `test/Dialect/ArmSME/vector-ops-to-llvm.mlir`,
`test/Dialect/ArmSME/tile-allocation-invalid.mlir`, `test/Conversion/ArmSMEToLLVM/arm-sme-to-llvm.mlir`, `test/Conversion/ArmSMEToLLVM/tile-spills-and-fills.mlir`, `test/Conversion/ArmSMEToLLVM/unsupported.mlir` when running with `MLIR_ENABLE_EXPENSIVE_PATTERN_API_CHECKS`.

### CODE REVIEW

File: mlir/lib/Dialect/ArmSME/Transforms/TileAllocation.cpp [hunk 1]
```diff
@@ -263,14 +260,26 @@ struct AssignTileIDsPattern
...
+    if (!tileIsInMemory)
+      setDiscardableIntAttr(kTilesInUseAttr, tilesInUse);
+    else {
+      setDiscardableIntAttr(kNextInMemoryTileIdAttr, *tileId + 1);
+    }
+    rewriter.updateRootInPlace(tileOp, [&] { tileOp.setTileId(tileIDAttr); });
+    for (auto *op : dependantOps) {
+      if (auto dependantTileOp = llvm::dyn_cast<ArmSMETileOpInterface>(op)) {
+        rewriter.updateRootInPlace(
+            dependantTileOp, [&]() { dependantTileOp.setTileId(tileIDAttr); });
```
Comment (MacDue) [inline mlir/lib/Dialect/ArmSME/Transforms/TileAllocation.cpp:282]: Just a clarification: This has to be done for an update of _any_ operation within a rewrite? If so, it's not immediately clear to me what "root" refers to in the API (and why it's not just `updateOpInPlace`)?

File: mlir/lib/Dialect/ArmSME/Transforms/TileAllocation.cpp [hunk 2]
```diff
@@ -263,14 +260,26 @@ struct AssignTileIDsPattern
...
-        dependantTileOp.setTileId(tileIDAttr);
+      }
+    }
+
+    // Rewrite IR.
+    if (!tileIsInMemory)
+      setDiscardableIntAttr(kTilesInUseAttr, tilesInUse);
+    else {
+      setDiscardableIntAttr(kNextInMemoryTileIdAttr, *tileId + 1);
+    }
```
Comment (MacDue) [inline mlir/lib/Dialect/ArmSME/Transforms/TileAllocation.cpp:277]: nit: 
```suggestion
    else
      setDiscardableIntAttr(kNextInMemoryTileIdAttr, *tileId + 1);
```

File: mlir/lib/Dialect/ArmSME/Transforms/TileAllocation.cpp (hunk 1 above)
Comment (MacDue) [inline mlir/lib/Dialect/ArmSME/Transforms/TileAllocation.cpp:282]: nit:
```suggestion
            dependantTileOp, [&] { dependantTileOp.setTileId(tileIDAttr); });
```

File: mlir/lib/Dialect/ArmSVE/Transforms/LegalizeVectorStorage.cpp [hunk 3]
```diff
@@ -106,7 +106,8 @@ struct RelaxScalableVectorAllocaAlignment
 
     // Set alignment based on the defaults for SVE vectors and predicates.
     unsigned aligment = vectorType.getElementType().isInteger(1) ? 2 : 16;
-    allocaOp.setAlignment(aligment);
+    rewriter.updateRootInPlace(allocaOp,
+                               [&]() { allocaOp.setAlignment(aligment); });
```
Comment (MacDue) [inline mlir/lib/Dialect/ArmSVE/Transforms/LegalizeVectorStorage.cpp:110]: nit:
```suggestion
                               [&] { allocaOp.setAlignment(aligment); });
```

File: mlir/lib/Dialect/ArmSME/Transforms/TileAllocation.cpp (hunk 1 above)
Comment (matthias-springer) [reply mlir/lib/Dialect/ArmSME/Transforms/TileAllocation.cpp]: Here's an attempt to improve the situation: #78260

**Review by MacDue — APPROVED** [review]
LGTM, thanks for the fixes!

**Pull request merged**